### DISCUSSION TITLE
Supporting Statechart concepts

### DISCUSSION THREAD

**Discussion by tmc** [terrastruct/d2#1128] [General]
Loving what you all are doing with D2!

I’d like to start a discussion around to approach introducing state chart concepts into D2.

In particular I’d love to find the right way to approach adding the ability to express Othogonal/Parallel states. https://statecharts.dev/glossary/parallel-state.html

PlantUML prior art: https://plantuml.com/state-diagram

**Comment by gavin-ts** [comment]
I wonder if we can support this by adding new shapes for initial, final, and condition?

e.g. this dot for initial:

![Screen Shot 2023-04-13 at 10 58 53 AM](https://user-images.githubusercontent.com/85081687/231845185-18bb8dd5-2191-47a1-bc85-5fa48d8d71f5.png)

For parallel states, I wonder if grid layouts #1122 could be used if we support a recursive mix of grid layouts and non-grid

**Reply by alixander** [reply]
why can't the dot just be styled though? shape circle, small width, black fill. it won't have that little curvy connection, but that's minor i think.

**Reply by alixander** [reply]
>  if we support a recursive mix of grid layouts and non-grid

yeah i think that would do the trick

**Reply by tmc** [reply]
The initial states are pretty important to the visual language of statecharts and it'd be unfortunate if those can't be naturally expressed. What might some possible paths be to get the nice little arrows? https://xstate.js.org/viz/ has some possible inspiration.

**Comment by tmc** [comment]
Relatedly, I opened https://github.com/terrastruct/text-to-diagram-site/issues/50 for the comparison site.
